# GTEx

This is an [R Markdown](http://rmarkdown.rstudio.com) Notebook. When you execute code within the notebook, the results appear beneath the code. 

Try executing this chunk by clicking the *Run* button within the chunk or by placing your cursor inside it and pressing *Cmd+Shift+Enter*. 

```{r}
# prediction based on brain expression
gtex_avg_expression = read_tsv("../data/raw_data/GTEx_Analysis_2016-01-15_v7_RNASeQCv1.1.8_gene_median_tpm.gct", skip = 2)

tissues_details <- read_csv("../data/tissue_combined_details.csv")

ensg_to_symbol <- readRDS("../cache/ensg_to_entrez_and_symbol_df.RDS")

protein_coding_sym <- read_tsv("../cache/coding_genes_in_gtex.tsv")
protein_coding_sym$protein_coding = T

ensg_to_symbol <- full_join(ensg_to_symbol, protein_coding_sym) %>%
  dplyr::rename("Description" = "symbol")

gtex_avg_expression_with_entrez <- full_join(ensg_to_symbol, gtex_avg_expression)

  
gtex_expression_count <- list()
gtex_expression_count$AllReported <- apply(gtex_avg_expression_with_entrez[,-c(1:8)],2, function(x) sum(x>0, na.rm = T))
gtex_expression_count$Protein_coding <- apply(gtex_avg_expression_with_entrez %>% filter(!is.na(protein_coding)) %>% select(-c(1:8))   ,2, function(x) sum(x>0, na.rm = T))
gtex_expression_count$avg_TPMover1 <- apply(gtex_avg_expression_with_entrez %>% filter(!is.na(protein_coding)) %>% select(-c(1:8))   ,2, function(x) sum(x>=1, na.rm = T))

#gtex_expression_count_df <- as.data.frame(do.call(rbind,gtex_expression_count)) %>% rownames_to_column(var = "Measure")
#gtex_expression_count_df<- pivot_longer(gtex_expression_count_df, cols = colnames(gtex_expression_count_df)[-1], values_to = "count", names_to = "tissue")


source("../functions/readdata_functions.R")
library(igraph)
g = process_graph_data("../data/network_edgelists/")
coex_nets <- names(g)[grepl("coex", names(g))]
nodes <- lapply(coex_nets, function(x) V(g[[x]])$name)
names(nodes) <- coex_nets

gtex_expression_count$Final_nodes <- sapply(nodes, length)


# check with version before the disp filter
###########

count_nodes_from_edgelist <- function(el_file){
  el <- read_tsv(paste0("../cache/temp/new_coex_by_group//", el_file), col_types = "ff")
  #nodes <- length(union(levels(el$A), levels(el$B)))
  nodes <- length(union(levels(el$gene1), levels(el$gene2)))
  edges <- nrow(el)
  return(c(nodes = nodes, edges = edges))
}

files <- list.files("../cache/temp/new_coex_by_group//")

full_nodes = list()
for(i in files){
  print(i)
  full_nodes[[i]] = count_nodes_from_edgelist(i)
}

gtex_expression_count$BeforeDisparity <- sapply(full_nodes, function(x) x['nodes'])
gtex_expression_count$BeforeDisparity = ifelse(gtex_expression_count$BeforeDisparity < gtex_expression_count$Final_nodes, gtex_expression_count$Final_nodes, gtex_expression_count$BeforeDisparity)
###########

gtex_expression_count_df <-reshape2:: melt(gtex_expression_count)
colnames(gtex_expression_count_df) = c("count", "Measure")

gtex_expression_count_df <- gtex_expression_count_df %>%
  mutate(Measure = factor(Measure, levels =  c("AllReported",  "Protein_coding",  "avg_TPMover1", "BeforeDisparity", "Final_nodes"), label = c("All transcripts", "Protein\ncoding", "Reference: \nTPM > 1", "Before\ndisp. filter", "Nodes in\nfinal networks")))


edge_count <- list()
edge_count[["Before\ndisp. filter"]] <- sapply(full_nodes, function(x) x['edges'])
edge_count[["Nodes in\nfinal networks"]] <- sapply(coex_nets, function(x) ecount(g[[x]]))
edge_count_df <-reshape2:: melt(edge_count)
colnames(edge_count_df) = c("count", "Measure")


# Plotting
########

colors <- c('#7fc97f','#beaed4','#fdc086','#ffff99','#386cb0')

p_nodes <- ggplot(gtex_expression_count_df, aes(x = Measure, y = count, fill = Measure)) + 
  geom_violin() + 
     geom_boxplot(width = 0.2, fill = "white") +
  guides(fill = F) + cowplot::theme_cowplot() + xlab("Criterion") + ylab("Count") + ggtitle("# Nodes") + scale_fill_manual(values = colors) 


p_edges <- ggplot(edge_count_df, aes(x = Measure, y = count, fill = Measure)) + 
  geom_violin() + 
   geom_boxplot(width = 0.2, fill = "white") +
  guides(fill = F) + cowplot::theme_cowplot() + xlab("Criterion") + ylab("Count") + ggtitle("# Edges") + scale_fill_manual(values = colors[4:5]) + scale_y_log10()


p_coex <- p_nodes +p_edges + plot_layout(widths = c(2, 1))

ggsave("../Figs/number_of_nodes_and_edges_coexpression_at_each_stages.pdf", p_coex, width = 6, height = 3, scale = 1.5)


p_coex  
```



```{r}
gtex_avg_expression_tissue <- list()

gtex_avg_expression_tissue[["Adipose (ADV)"]] <- gtex_avg_expression_with_entrez  %>% filter(protein_coding, is.na(notfound)) %>% mutate(inTissue = (Description %in% nodes$coex_ADV)) %>% select(Description, inTissue,`Adipose - Subcutaneous`) %>% rename(expression = colnames(.)[3])
gtex_avg_expression_tissue[["Heart (HRA)"]] <- gtex_avg_expression_with_entrez  %>% filter(protein_coding, is.na(notfound)) %>% mutate(inTissue = (Description %in% nodes$coex_HRA)) %>% select(Description, inTissue,`Heart - Atrial Appendage`) %>% rename(expression = colnames(.)[3])
gtex_avg_expression_tissue[["Lung (LNG)"]] <- gtex_avg_expression_with_entrez  %>% filter(protein_coding, is.na(notfound)) %>% mutate(inTissue = (Description %in% nodes$coex_LNG)) %>% select(Description, inTissue, Lung) %>% rename(expression = colnames(.)[3])
gtex_avg_expression_tissue[["Ovary (OVR)"]] <- gtex_avg_expression_with_entrez  %>% filter(protein_coding, is.na(notfound)) %>% mutate(inTissue = (Description %in% nodes$coex_OVR)) %>% select(Description, inTissue, Ovary) %>% rename(expression = colnames(.)[3])
gtex_avg_expression_tissue[["Whole Blood (WBL)"]] <- gtex_avg_expression_with_entrez  %>% filter(protein_coding, is.na(notfound)) %>% mutate(inTissue = (Description %in% nodes$coex_WBL)) %>% select(Description, inTissue, `Whole Blood`) %>% rename(expression = colnames(.)[3])
gtex_avg_expression_tissue[["Brain (BRO)"]] <- gtex_avg_expression_with_entrez  %>% filter(protein_coding, is.na(notfound)) %>% mutate(inTissue = (Description %in% nodes$coex_BRO)) %>% select(Description, inTissue, `Brain - Caudate (basal ganglia)`) %>% rename(expression = colnames(.)[3])
gtex_avg_expression_tissue[["Pituitary (PIT)"]] <- gtex_avg_expression_with_entrez  %>% filter(protein_coding, is.na(notfound)) %>% mutate(inTissue = (Description %in% nodes$coex_PIT)) %>% select(Description, inTissue, Pituitary) %>% rename(expression = colnames(.)[3])
gtex_avg_expression_tissue[["Skin (SKN)"]] <- gtex_avg_expression_with_entrez  %>% filter(protein_coding, is.na(notfound)) %>% mutate(inTissue = (Description %in% nodes$coex_SKN)) %>% select(Description, inTissue, `Skin - Not Sun Exposed (Suprapubic)`) %>% rename(expression = colnames(.)[3])
gtex_avg_expression_tissue[["Kindney (KDN)"]] <- gtex_avg_expression_with_entrez  %>% filter(protein_coding, is.na(notfound)) %>% mutate(inTissue = (Description %in% nodes$coex_KDN)) %>% select(Description, inTissue, `Kidney - Cortex`) %>% rename(expression = colnames(.)[3])
gtex_avg_expression_tissue[["Thyroid (THY)"]] <- gtex_avg_expression_with_entrez  %>% filter(protein_coding, is.na(notfound)) %>% mutate(inTissue = (Description %in% nodes$coex_THY)) %>% select(Description, inTissue, Thyroid) %>% rename(expression = colnames(.)[3])
gtex_avg_expression_tissue[["Esophagus (GEJ)"]] <- gtex_avg_expression_with_entrez  %>% filter(protein_coding, is.na(notfound)) %>% mutate(inTissue = (Description %in% nodes$coex_GEJ)) %>% select(Description, inTissue, `Esophagus - Gastroesophageal Junction`) %>% rename(expression = colnames(.)[3])
gtex_avg_expression_tissue[["Salivary Gland (MSG)"]] <- gtex_avg_expression_with_entrez  %>% filter(protein_coding, is.na(notfound)) %>% mutate(inTissue = (Description %in% nodes$coex_MSG)) %>% select(Description, inTissue, `Minor Salivary Gland`) %>% rename(expression = colnames(.)[3])


gtex_avg_expression_tissue_df <- bind_rows(gtex_avg_expression_tissue, .id = "Tissue") %>% filter(expression > 0, !is.na(expression))


p <- ggplot(gtex_avg_expression_tissue_df, aes(x = log2(expression+1), fill = inTissue, col = inTissue)) + geom_density(alpha = 0.1) + facet_wrap(~Tissue, nrow = 2, scales = "free") + cowplot::theme_cowplot() + xlab("Expression value (TPM)") + ggtitle("Average expression level")

ggsave("../Figs/example_expression_level_all_vs_in_network.pdf", p, width = 12, height = 4, scale = 1.2)

 p
```


Interactions exists in PPI but no co-expression?
```{r}
genes = c(
BRCA1 = "ENSG00000012048",
GATA3 = "ENSG00000107485",
GABPB2 = "ENSG00000143458",
MSH6 = "ENSG00000095002")


a <- readRDS("../cache/temp/expression_merged_tissue_group.RDS")
#load("../cache/temp/E-MTAB-5214-atlasExperimentSummary.Rdata")
#dat <- a$BST[genes,] %>% t() %>% as_tibble() %>% pivot_longer(., cols = colnames(.), names_to = "gene", values_to = "expression") %>%
#  mutate(gene = factor(gene, levels = genes, labels = names(genes)))

dat1 <- a$BST[genes[1:2],] %>% t() %>% as_tibble() 
colnames(dat1) = c("gene1", "gene2")

p1 <- ggplot(dat1, aes(x = gene1, y = gene2)) + geom_point(col = "grey25", alpha = 0.6) + scale_x_log10() + scale_y_log10() + ggtitle("BRCA1-GATA3", subtitle = paste0("cor: ", round(cor(dat1$gene1, dat1$gene2, method = "pearson"), 3))) + theme_cowplot() + xlab("BRCA1 expression") + ylab("GATA3 expression")+ theme(axis.text.x = element_blank(), axis.text.y = element_blank(), axis.ticks = element_blank())



dat2 <- a$BST[genes[c(1,3)],] %>% t() %>% as_tibble() 
colnames(dat2) = c("gene1", "gene2")

p2 <- ggplot(dat2, aes(x = gene1, y = gene2)) + geom_point(col = "grey25", alpha = 0.6) + scale_x_log10() + scale_y_log10() + ggtitle("BRCA1-GABPB2", subtitle = paste0("cor: ", round(cor(dat2$gene1, dat2$gene2, method = "pearson"), 3))) + theme_cowplot()+ xlab("BRCA1 expression") + ylab("GABPB2 expression") + theme(axis.text.x = element_blank(), axis.text.y = element_blank(), axis.ticks = element_blank())



dat3 <- a$BST[genes[c(1,4)],] %>% t() %>% as_tibble() 
colnames(dat3) = c("gene1", "gene2")

p3 <- ggplot(dat3, aes(x = gene1, y = gene2)) + geom_point(col = "grey25", alpha = 0.6) + scale_x_log10() + scale_y_log10() + ggtitle("BRCA1-MSH6", subtitle = paste0("cor: ", round(cor(dat3$gene1, dat3$gene2, method = "pearson"), 3))) + theme_cowplot() + xlab("BRCA1 expression") + ylab("MSH6 expression")+ theme(axis.text.x = element_blank(), axis.text.y = element_blank(), axis.ticks = element_blank())



library(patchwork)
p_gene <- p1 + p2


p_gene

ggsave("../Figs/BRCA1_expression_level_GATA3(foundinPPI)_GABP2(inBreastCoex).pdf", plot = p_gene, width = 3, height = 1.6, scale = 2)
```


## Some important genes could be removed by using a hard cutoff

```{r}
dat <- gtex
```



Add a new chunk by clicking the *Insert Chunk* button on the toolbar or by pressing *Cmd+Option+I*.

When you save the notebook, an HTML file containing the code and output will be saved alongside it (click the *Preview* button or press *Cmd+Shift+K* to preview the HTML file). 
 
The preview shows you a rendered HTML copy of the contents of the editor. Consequently, unlike *Knit*, *Preview* does not run any R code chunks. Instead, the output of the chunk when it was last run in the editor is displayed.

```{r}

names(nodes) <- coex_nets

table(nodes$coex_ADV %in% 
(gtex_avg_expression_with_entrez %>% filter(!is.na(protein_coding), `Adipose - Visceral (Omentum)` > 0.1) %>% pull(Description)))
```



```{r}
# correlation plot


cor_HRV <- readRDS("../cache/temp/coex_group_pairwise_val_HRV.RDS")

cor_index <- readRDS("../cache/temp/coex_group_pairwise_gene_index.RDS")

cor_HRV <- cor_HRV %>% filter(!is.na(HRV))
cor_HRV_cut <- cut(cor_HRV$HRV, breaks = seq(0,1,0.05))

cor_count <- table(cor_HRV_cut) %>% as_tibble()
cor_count$label <- seq(0,1,0.05)[-1]

ggplot(cor_count, aes(x=label, y = n)) + geom_col()


cor_index$cor_HRV <- cor_HRV$HRV
cor_index <- cor_index %>% filter(!is.na(cor_HRV))

el_HRV <- as_data_frame(g$coex_HRV)
el_HRV$pass <- T

#cor_index_merge <- left_join(cor_index, el_HRV, by = c(A="from", B="to"))
cor_index_merge <- left_join(cor_index, el_HRV, by = c(B="from", A="to"))
cor_index_merge$pass[is.na(cor_index_merge$pass)]  = F
cor_index_merge$group <- cut(cor_index_merge$cor_HRV, breaks = seq(0,1,0.05), labels = seq(0,1,0.05)[-1], ordered_result = T)

cor_index_merge_group_count <- count(cor_index_merge, group, pass)

ggplot(cor_index_merge_group_count %>% filter(group %in% seq(0.75,1,0.05)), aes(x = group, y = n, fill = pass)) + geom_col(position = "dodge") + scale_y_log10()
```

```{r}
BRCA1_expression <-gtex_avg_expression_with_entrez  %>%  filter(protein_coding) %>% select(Description, `Breast - Mammary Tissue`, Ovary)

BRCA1_percentile <- BRCA1_expression %>%
  mutate(rank_breast = rank(- `Breast - Mammary Tissue`)/sum(gtex_avg_expression_with_entrez$`Breast - Mammary Tissue`>0, na.rm = T),
         rank_ovary = rank(- Ovary)/sum(gtex_avg_expression_with_entrez$Ovary>0, na.rm = T)
         ) %>%
  filter(Description == "BRCA1")
```

